src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs

```diff
@@ -0,0 +1,332 @@
+            }
+
+            var apiDescription = new ApiDescription
+            {
+                HttpMethod = httpMethod,
+                RelativePath = routeEndpoint.RoutePattern.RawText?.TrimStart('/'),
+                ActionDescriptor = new ActionDescriptor
+                {
+                    RouteValues =
+                    {
+                        ["controller"] = controllerName,
+                    },
+                },
+            };
+
+            var hasJsonBody = false;
+
+            foreach (var parameter in methodInfo.GetParameters())
+            {
+                var parameterDescription = CreateApiParameterDescription(parameter, routeEndpoint.RoutePattern);
+
+                if (parameterDescription.Source == BindingSource.Body)
+                {
+                    hasJsonBody = true;
+                }
+
+                apiDescription.ParameterDescriptions.Add(parameterDescription);
+            }
+
+            AddSupportedRequestFormats(apiDescription.SupportedRequestFormats, hasJsonBody, routeEndpoint.Metadata);
+            AddSupportedResponseTypes(apiDescription.SupportedResponseTypes, methodInfo.ReturnType, routeEndpoint.Metadata);
+
+            return apiDescription;
+        }
+
+        private ApiParameterDescription CreateApiParameterDescription(ParameterInfo parameter, RoutePattern pattern)
+        {
+            var (source, name) = GetBindingSourceAndName(parameter, pattern);
+
+            return new ApiParameterDescription
+            {
+                Name = name,
+                ModelMetadata = CreateModelMetadata(parameter.ParameterType),
+                Source = source,
+                DefaultValue = parameter.DefaultValue,
+                Type = parameter.ParameterType,
+            };
+        }
+
+        // TODO: Share more of this logic with RequestDelegateFactory.CreateArgument(...) using RequestDelegateFactoryUtilities
+        // which is shared source.
+        private (BindingSource, string) GetBindingSourceAndName(ParameterInfo parameter, RoutePattern pattern)
+        {
+            var attributes = parameter.GetCustomAttributes();
+
+            if (attributes.OfType<IFromRouteMetadata>().FirstOrDefault() is { } routeAttribute)
+            {
+                return (BindingSource.Path, routeAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (attributes.OfType<IFromQueryMetadata>().FirstOrDefault() is { } queryAttribute)
+            {
+                return (BindingSource.Query, queryAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (attributes.OfType<IFromHeaderMetadata>().FirstOrDefault() is { } headerAttribute)
+            {
+                return (BindingSource.Header, headerAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromBodyMetadata).IsAssignableFrom(a.AttributeType)))
```

Comment:
Does `OfType` not work for this?

Reply:
Yes. Technically Any is a bit better because it avoids the iterator, but it's not like this is super high perf code. If we change this we should do the same in RequestDelegateFactory. Maybe in a later cleanup.